### PR TITLE
Dynamic Filtering (both Partition Pruning and Row filtering)

### PR DESCRIPTION
**Pull request by amoghmargoor**
We have  decided to break this PR into multiple chunks for ease of reviewing.

### CODE REVIEW

**Comment by amoghmargoor** [comment]
@sopel39 Hi Karol, Can you help getting this reviewed ? UTs are fixed except for the docker test which I think are failing even on master. We are seeing good performance benefit due to this, which I would update soon.

**Comment by amoghmargoor** [comment]
chart_final

Attaching the improvements in tpcds queries 7,15,31,34,46,71 and 79. It is run on TPCDS scale 1000 on AWS with data being on S3 in ORC format. It is non-partitioned data so most of the benefit is due to Dynamic Row filtering operators. We ported this to presto0.180 offered at Qubole before running this benchmark. We are seeing avg reduction in execution time of 47% in the above queries, max reduction of 72 % and min reduction of 15%.

cc @sopel39 @dain @losipiuk @highker

**Comment by sopel39** [comment]
I have looked briefly at how `DynamicFilter*` is structured. You have introduced various variants of that filter depending on use case:
```
* DynamicFilterAndProjectOperator
* DynamicFilterSourceOperator
* DynamicScanFilterAndProjectOperator
``` 
with bunch of surrounding classes. `DynamicFilter*` operators extends basic version of operators (e.g: `FilterAndProjectOperator`). Generally I think it's not the approach we would like to follow because it's overly complicated and hard to maintain.

What I would propose instead is to implement dynamic filtering as a dedicated `PageFilter`. This would remove the need for creating new matrix of operators. In planner we could either:
* we could have logical expression like: `dynamic_filter(col_a,col_b,...,"dynamicFilterName")` which would be treated as a normal expression. Such `dynamic_filter` expression would be pushed down using normal `PredicatePushdown` mechanism and it would be converted to dedicated `PageFilter` in `PageFunctionCompiler`. `dynamic_filter` should be always part of top level `conjunction`. 
* still have `DynamicFilter` node, but such node would only be logical and during execution planning it would be converted to either `FilterAndProjectOperator` or it would be part of `ScanFilterAndProjectOperator`. 

I like the former approach more since it reuses existing mechanisms (e.g: `PredicatePushdown`), and does not require any new dedicated plan nodes. However, I wonder if such `dynamic_filter` won't be pushed above `join` in the future (e.g: by CBO) and if it will always be part of top level conjunction (e.g: is it really like a normal expression).
My intuition is that explicit `DynamicFilter` node approach might be more solid.

Your opinion: @martint @kokosing @findepi ?

**Comment by sopel39** [comment]
It also seems that dynamic filters summary is distributed though coordinator (please correct me if I am wrong). This might cause high network load on coordinator. For instance, if there are 100 cluster nodes and each node sends 1MB of data, then coordinator has to distribute 10GB of filter data per query per join. The best would be to have a stage with multiple outputs, but that's probably very difficult in current code base. Simpler solution would be for `DynamicFilterService` on coordinator to redirect filter pull queries to appropriate cluster nodes.
@dain ?

**Comment by amoghmargoor** [comment]
@sopel39 Thanks for the review.
Yes, we do use this version. We have limited cardinality of TupleDomain being sent to DF service running on coordinator which mostly takes care of data heavy build side in practise by reducing Tuple Domain to true after it hits limit (1000 values). Similarly, we can as well put limit on cardinality on final merged tuple domain too so that data being distributed to consumer side is not high. However in practise we have not seen such high data being sent to coordinator as the first limit generally is enough to reduce TupleDomains to true for data heavy build side. We can take that up if such issues are seen in practise. But that reminds me of something else: How do we get the coordinator IP and port ? Currently this PR is using Discovery IP which is different than what we do in production. In our stack we add the coordinator IP to config before bringing up presto daemon. I was not sure what is the best way to fetch it here. 

Regarding the earlier comment, just to summarise the action item:
1. Implement new `PageFilter`.  Would this also be applicable for cursor processor as such processors are being used in `ScanFilterAndProjectOperator` ?
2. Implement `DynamicFilter`. Currently FilterNode does the same thing. It can contain Expressions having DynamicFilter and those can be pushed down etc. As you pointed out there is a issue if DF are pushed above joins etc due to CBO. There is new rule added `RemoveUnsupportedDynamicFilters` which should take care of removing all DF which have not been pushed till the the table scan. 
3. Remove new DF operators extending `ScanFilterAndProjectOperator` and `FilterAndProjectOperator`. Sure, will remove new operators. So in the existing operator, after DF  gets resolved into static filter,  just recreating new processor (either cursor or page) or new `PageFilter` (if it is also applicable to cursor processor) be enough. Until DF is not resolved old processor or PageFilter would be used and processing would not be blocked. Does that sound reasonable ?
Let me know if I am missing any other details of the comments and really appreciate all your help.

**Comment by sopel39** [comment]
> Domain to true after it hits limit (1000 values).

So basically it currently works for very small tables, right? I would imagine in the future the natural way to progress would be to use bloom filters for large tables, which can be quite large if we want to have reasonable false positives rate. I guess using coordinator for multi-casting could be good for now for small tables, but I would rather have 3rd opinion on that (e.g: @dain @martint)?

> Implement DynamicFilter

My original idea was to make `DynamicFilter` just a logical plan node which would be translated into `FilterAndProjectOperator` or incorporated into `ScanFilterAndProjectOperator` as dedicated `PageFilter`. However, I have seen that this PR introduces `DeferredSymbolReference` which I believe can be used only in equation context and essentially represents dynamic filter.
I think `DeferredSymbolReference` should actually be refactored into `DynamicFilterExpression` 
that would contain:
* `columns`: list of `SymbolReferences` - columns which are filtered
* `source`: dynamic filter source name

It would basically model expression like: `dynamic_filter(co1, col2, ..., "source")`. `DeferredSymbolReference` is not the best abstraction since it suggest it can be used as normal expression, but in fact it can only be used in equation context. In any case we don't need any new physical operators since filtering would be implemented as `PageFilter`.

>  Would this also be applicable for cursor processor as such processors are being used in ScanFilterAndProjectOperator

I don't think we should bother with cursor processor for now. This is because it reads files row by row and we don't get any benefit of lazy load when using dynamic filtering. If we need then for cursor based sources we can have a separate `FilterAndProjectOperator` above `ScanFilterAndProjectOperator` that would only do page based dynamic filtering.

**Comment by sopel39** [comment]
> Domain to true after it hits limit (1000 values)

I think it would be better to use size limit instead of positions limit. It's not unreasonable to expect that join columns would consist of particularly long `VARCHARS`.

**Comment by amoghmargoor** [comment]
@sopel39 I have removed new operators and created DynamicPageFilter now. I still need to refactor DeferredSymbolReference into DynamicFilterExpression

**Comment by amoghmargoor** [comment]
@sopel39 Do new changes look fine ? If yes then I will plan to rebase it with current master.

**Comment by kokosing** [comment]
@amoghmargoor To maintain great relationship of Starburstdata with Qubole, I would like to cooperate with you to merge this PR to master.

Would you mind to update the desing doc to reflect the current approach: https://docs.google.com/document/d/1NYhEqYfmOPUqYFVzmOWvIXds5reFlaAi7z-zeJWqSLE/edit#. Please correct me if I am wrong, currently there is no such thing like `DynamicFilter` node, instead an `dynamic filter scalar function` expression is used in `FilterNode`. This is a big conceptual change. What about `DynamicFilterSource`?

Also please do the rebase and squash the fixup commits.

**Comment by kokosing** [comment]
> How do we get the coordinator IP and port ? Currently this PR is using Discovery IP

We could start with it, in future we could set needed information into `DynamicFilterSource` operator so it would know where to send information.

>> Domain to true after it hits limit (1000 values)

> I think it would be better to use size limit instead of positions limit. It's not unreasonable to expect that join columns would consist of particularly long VARCHARS.

I think to begin with we could have a limit like that and limit tuple domain to support only `Long`-based columns.

> > Would this also be applicable for cursor processor as such processors are being used in ScanFilterAndProjectOperator

> I don't think we should bother with cursor processor for now. This is because it reads data row by row and we don't get any benefit of lazy load when using dynamic filtering. If we need then for cursor based sources we can have a separate FilterAndProjectOperator above ScanFilterAndProjectOperator that would only do page based dynamic filtering.

I agree with @sopel39 
---

I also like approach with `dynamic_filter(col_a,col_b,...,"dynamicFilterName")` much better. Also, I think it will be much easier to have a `DynamicFilterSource` node with matching `dynamicFilterName`. I would avoid changing `JoinNode` and join execution.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 1]
```diff
@@ -83,6 +83,7 @@
     public static final String FORCE_SINGLE_NODE_OUTPUT = "force_single_node_output";
     public static final String FILTER_AND_PROJECT_MIN_OUTPUT_PAGE_SIZE = "filter_and_project_min_output_page_size";
     public static final String FILTER_AND_PROJECT_MIN_OUTPUT_PAGE_ROW_COUNT = "filter_and_project_min_output_page_row_count";
+    public static final String DYNAMIC_PARTITION_PRUNING = "dynamic_partition_pruning";
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: If we go `dynamic_filter(...)` expression way, then whole dynamic filtering pushdown won't be needed because it will be automatically supported by `PredicatePushdown`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java [hunk 2]
```diff
@@ -57,6 +57,7 @@
     private final Optional<Symbol> leftHashSymbol;
     private final Optional<Symbol> rightHashSymbol;
     private final Optional<DistributionType> distributionType;
+    private final Assignments dynamicFilterAssignments;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java]: I don't think we need `Assignments` here, do we?
I think we just need to store:
```
List<Symbol> dynamicFilterSymbols;
```
Motivation is: once we create `dynamic_filter` expressions for a given join, then we cannot change order of symbols in dynamic filtering tuples.

**Comment by amoghmargoor** [comment]
@sopel39  @kokosing - I have rebased commits with latest master ... I will start working on rest of the comments and hopefully finish it in couple of days. Thanks.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java]: We might want to keep `Assignments` here, although with a slightly different meaning (basically for validating plan correctness). I will write more tomorrow.

**Comment by sopel39** [comment]
I see you squashed all commits. Would it be possible to rebase without squashing all commits, but only fixups?

**Comment by kokosing** [comment]
> Also please do the rebase and squash the fixup commits.

@amoghmargoor I am very sorry, it looks like I was not clear enough. Can you restore the previous commits layout with the original authorship? However please squash all the commits which exists only due some mistakes in preceding commits in this PR. Such fixup commits should squashed with the commits they fix. The idea is to have all the commits compilable and testable as well as to make each commit to be a single logical change (step) which brings codebase a bit closer of the goal of this PR.

As a result you should not see any commits like `!fixup:...`, `Fix compilation`, `Fix checkstyle` or `Fix tests`.

This is not easy feature to understand and review. We all want to make this code as best as possible. In order to reduce the maintenance effort, to make sure it fits perfectly into the design of Presto and to make it easily extendable in the future. Having 4k of lines per single commit makes it difficult to review, test and merge. For example in case of bug it is easier and better to revert only single small buggy commit than the whole feature. Also having multiple commits makes it easier to merge as we could do it partially. Having 400 commits would not be also a good, as there is too much effort to maintain the PR. We need to find a good tradeoff. I believe we will reach that during the process of review.

Please also avoid changing the same code (change) twice in a single PR. Simple example, if you rename A to B, and then in the same PR you rename B to C, please squash such commits to remain single commit where A is renamed to C.

Keeping the original authorship is to us very important. @maciejgrzybek and @arhimondr are not currently actively involved in project, but they committed enormous effort in this work, it would very impolite to erase the proof they contributed to this feature. Also notice that keeping original authorship makes it clear that this feature is the effect of joint effort of you (Qubole), them (TD), us (Starburst) and obviously as always (FB).

**Comment by amoghmargoor** [comment]
@kokosing Looks like I didn't understand the comment well enough and ended up squashing all commits together ....Really sorry about that ... Let me try to restore original PR commits atleast.

File: presto-main/src/main/java/com/facebook/presto/sql/DynamicFilter.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.DynamicFilterUtils.isDynamicFilter;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilter
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/DynamicFilter.java]: See my other comment: https://github.com/prestodb/presto/pull/9453#discussion_r168467073

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 4]
```diff
@@ -619,4 +625,9 @@ public static int getFilterAndProjectMinOutputPageRowCount(Session session)
     {
         return session.getSystemProperty(FILTER_AND_PROJECT_MIN_OUTPUT_PAGE_ROW_COUNT, Integer.class);
     }
+
+    public static boolean isDynamicPartitionPruningEnabled(Session session)
```
[the remark this answers is not in the document]
Comment (amoghmargoor) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: it is being used but not as a flag to just PartitionPruning; instead being used as flag for entire DF. I need to rename it.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 1 above)
Comment (amoghmargoor) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: same as above comment. this flag is being used for enabling entire DF feature. needs rename.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java (hunk 2 above)
Comment (amoghmargoor) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java]: Let me go through this comment in detail and revert back.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 5]
```diff
@@ -434,7 +434,7 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
             }
```
Comment (kokosing) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Please fix the commit message. It is not a revert, at that point of time it was ok to remove this field. There is nothing wrong with that commit so it does not have to be reverted. You just `Add hiveType field to HivePartitionKey`.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java [hunk 6]
```diff
@@ -76,6 +77,12 @@
...
         extends BasePlanTest
 {
+    TestLogicalPlanner()
+    {
+        // in order to test testUncorrelatedSubqueries with Dynamic Filtering, enable it
+        super(ImmutableMap.of(DYNAMIC_PARTITION_PRUNING, "true"));
+    }
+
     @Test
     public void testDistinctLimitOverInequalityJoin()
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java]: Please add tests like that to check that `DynamicFilter` related things are set correctly for supported queries. They should be added in class like `TestDynamicFilter` in this package.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilterExpression
+        extends Expression
+{
+    private final String sourceId;
+    private final ComparisonExpressionType type;
+    private final SymbolReference probeSymbolReference;
```
Comment (sopel39) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: This should be `Expression probeExpression`. Otherwise `DynamicFilterExpression` cannot be inlined with projection that produces `probeSymbolReference`.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    private final String dfSymbol;
```
Comment (sopel39) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: I would rather have an `Assignments` here, e.g: `dynamic_filter_sym -> probe_expression`. 

This is because with many join conditions, e.g: `probe_sym1=build_sym1 AND probe_sym2=build_sym2` you can implement dynamic filtering efficiently as hash filtering on column tuples instead of hash filtering on each column one by one.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 5 above)
Comment (amoghmargoor) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Done. Changed the commit message.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 9]
```diff
@@ -85,6 +85,7 @@
     public static final String FORCE_SINGLE_NODE_OUTPUT = "force_single_node_output";
     public static final String FILTER_AND_PROJECT_MIN_OUTPUT_PAGE_SIZE = "filter_and_project_min_output_page_size";
     public static final String FILTER_AND_PROJECT_MIN_OUTPUT_PAGE_ROW_COUNT = "filter_and_project_min_output_page_row_count";
+    public static final String DYNAMIC_FILTERING = "dynamic_filtering";
```
Comment (amoghmargoor) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: @kokosing renamed it to Dynamic filtering as suggested in review comments.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilterExpression
+        extends Expression
+{
+    private final String sourceId;
+    private final ComparisonExpressionType type;
+    private final Expression probeExpression;
```
Comment (amoghmargoor) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: @sopel39 Changed it to probeExpression instead of probeSymbolReference earlier. Other feedback was to use  `Assignments` - Assignments are defined in module presto-main and are not accessible in presto-parser.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestDynamicFilter.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.equiJoinClause;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.filter;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.join;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.node;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.semiJoin;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.tableScan;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+
+public class TestDynamicFilter
```
Comment (amoghmargoor) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestDynamicFilter.java]: @kokosing Added TestDynamicFilter as suggested in review comment for testing LogicalPlanning with DynamicFiltering on.

**Comment by amoghmargoor** [comment]
@kokosing @sopel39 I have taken care of all your comments. I will rebase and break this PR into 2 as suggested.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java (hunk 6 above)
Comment (amoghmargoor) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java]: Done.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java (hunk 7 above)
Comment (amoghmargoor) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: Done.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java (hunk 8 above)
Comment (amoghmargoor) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: Assignments not accessible here.

File: presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+                        BOOLEAN,
+                        rowExpression.get(),
+                        staticFilter.get());
+                resultExpression = Optional.of(combinedExpression);
+                globalState = DynamicProcessorState.DONE;
+                triggerCallback(combinedExpression);
+            }
+            catch (Exception e) {
+                globalState = DynamicProcessorState.FAILED;
+                throw e;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java]: will this exception be reported or logged?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 13]
```diff
@@ -1156,12 +1187,19 @@ private PhysicalOperation visitScanFilterAndProject(
...
-
                     return new PhysicalOperation(operatorFactory, outputMappings, groupEnumerable ? GROUPED_EXECUTION : UNGROUPED_EXECUTION);
                 }
                 else {
-                    Supplier<PageProcessor> pageProcessor = expressionCompiler.compilePageProcessor(translatedFilter, translatedProjections, Optional.of(context.getStageId() + "_" + planNodeId));
-
+                    Supplier<PageProcessor> pageProcessor;
+                    if (SystemSessionProperties.isDynamicFilteringEnabled(context.getSession()) && dynamicFilters.isPresent() && !dynamicFilters.get().isEmpty()) {
+                        RowExpressionConverter converter = new RowExpressionConverter(sourceLayout, expressionTypes, metadata.getFunctionRegistry(), metadata.getTypeManager(), session);
+                        DynamicFilterClient client = dynamicFilterClientSupplier.createClient(null, null, -1, -1, converter.getTypeManager());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: `null, null, -1, -1`. Remove those parameters from `createClient` method if unused?

File: presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+
+    public void collectDynamicSummaryAsync()
+    {
+        CompletableFuture.runAsync(this::collectDynamicSummary, es).whenComplete((value, exception) ->
+        {
+            this.cleanUp();
+        });
+    }
+
+    public void cleanUp()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java]: private?

File: presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Signatures.logicalExpressionSignature;
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilterCollector
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java]: `DynamicFilterCollector` seems not to be closed when `ScanFilterAndProject` operator closes. Can it query master after the `SFP` finishes?

File: presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+    }
+
+    public boolean isDone()
+    {
+        return globalState == DynamicProcessorState.DONE;
+    }
+
+    public interface DFCollectorCallback
+    {
+        void onSuccess(RowExpression result);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java]: Would it be possible to model collector as a service with method:
```
ListenableFuture<RowExpression> collectDFFilter(DynamicFilter dynamicFilter)
```
It would be more consistent with how asynchronous events are modeled in Presto. Such future could be cancelled when query waiting for it finishes early.

**Comment by losipiuk** [comment]
Hi! @amoghmargoor 

@sopel39 asked me to take a look at the PR and I must say that it is really hard to go through it all. It is quite big and same parts of codebase are touched in multiple commits so it is kinda hard to follow what is going on. My feeling is it would be better to abandon this PR as a whole. And just keep the code on some working branch. Then we could cut out concise parts of this working branch as separate small PRs.
IMO those PRs do not have to add user visible functionality. We must just make sure that they do not break Presto build or add any performance/functional degradetations.

That way we will probably end up with not so clean commit history as we could have, if we prepared everything on single PR branch and merged all together. 
But hopefully we will be able to iterate on this feature more effectively, and have it merged faster that way.

Here is the proposal what seem to be possible candidates for separate PRs extracted out of this one. Possibly some may be merged together if you see that would work better.
I listed commits from current PR which seem relevant by each bullet point. I assume some work need to be done to extract to appropriate bits from them. 
Note that some commits are listed more than once if they contain bits relevant to work items which can be separate PRs.

* Introduce config feature flag (disabled by default) which will guard new dynamic-filtering related code. 
* Introduce collecting `DynamicFilterSummary` on build side of Join.
  * ac6e9b44c6 - Introduce DynamicFilterSummary
  * 5717486046 - Collect DynamicFilterSummary before HashBuilder
* Introduce `DynamicFilterExpression` and optimizer rule which adds it in Filter below probe side of INNER join. 
  * 5622cb0a9d - Add DynamicFilter
  * fdc44e0450 - Add ApplyDynamicFilters
* Introduce a rule for pruning out stale `DynamicFilterExpression` for joins which are no longer in the plan. Or which are in unsupported place in plan.
  * fdc44e0450 - Add ApplyDynamicFilters
* Expose endpoint on coordinator for collecting `DynamicFilterSummaries` from worker nodes
  * f5c5984cf0 - Add DynamicFilterResource
* Feed coordinator with `DynamicFilterSummaries` from worker nodes.
  * f5c5984cf0 - Add DynamicFilterResource
  * 64ccbd35bf - Add InMemoryDynamicFilterClient
* Use `DynamicFilterSummaries` collected on coordinator to filter out splits in `SplitManager`.
  * b6d5318f80 - Register tasks in DynamicFilterService
  * f70525289d - Add DynamicFilterDescription
  * 9bae02f24d - Pass Future with DynamicFilterDescription to the SplitManager
  * 03fa054de4 - Add dynamic partition pruning to Hive connector
* Enable dynamic partitioning by default, or drop feature flag?
 
I did not go through your latter commits. Some of them probably also need to make it to one of above suggested PRs.

It seems that these two can go (at least mostly) away:
* fb9729a5e2 - Implement dynamic filtering pushdown 
* ef7efd90dc - Add DynamicFilterAssignments to JoinNode
Correct me if I am wrong - but I do not see `DynamicFilterAssignments` is needed if `DynamicFilter` is standard expression.

Here we have split filtering working. Now we can extend it with `TableScan` level filtering. I did not go through that part.

@amoghmargoor, @sopel39, @kokosing, @findepi, @arhimondr please let me know what you think about this. Does it make some sense?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class DynamicFilterExpression
+        extends Expression
+{
+    private final String sourceId;
+    private final ComparisonExpressionType type;
```
Comment (sopel39) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: I think we can remove `ComparisonExpressionType`. It's probably exotic case anyway and if we improve dynamic filtering to use Bloom filters, then this won't be supported.

File: presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java (hunk 15 above)
Comment (amoghmargoor) [reply presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java]: yeah currently it can ... it does close itself after retrying all attempts but it doesn't know if SFP finished. basically there is no close function in PageFilter that can be propagated to collector.

File: presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java (hunk 14 above)
Comment (amoghmargoor) [reply presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java]: done.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java (hunk 13 above)
Comment (amoghmargoor) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: done.

File: presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java (hunk 12 above)
Comment (amoghmargoor) [reply presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java]: no ...but  i added a log for it ... it will be caught upstream though.

File: presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java [hunk 18]
```diff
@@ -231,14 +231,24 @@ public Page getOutput()
...
+            return processPageSource(getPageProcessor());
         }
         else {
-            return processColumnSource();
+            return processColumnSource(getCursorProcessor());
         }
     }
 
-    private Page processColumnSource()
+    protected PageProcessor getPageProcessor()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java]: this is no longer needed, right?

File: presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java (hunk 15 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java]: Maybe it could be closed on `SFP::close()`. Alternatively, we could add `close` method to `PageFilter`.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java (hunk 17 above)
Comment (sopel39) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: ping

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java (hunk 10 above)
Comment (sopel39) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: > Other feedback was to use Assignments - Assignments are defined in module presto-main and are not accessible in presto-parser.

We could use a map here: `Map<Expression, String> probeExpressions`. However, I think it's fine for now. Most importantly, we got rid of deffered expressions, which were algebraically invalid concept.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java (hunk 17 above)
Comment (amoghmargoor) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/DynamicFilterExpression.java]: sure ... removing this too.
